### PR TITLE
Optimisation/ai spawner pooling

### PR DESCRIPTION
**Pull request by TJ-Smits**
Implemented a Generic Object Pooler for optimisation usage and implemented it for the AI spawning

### CODE REVIEW

File: behind-the-tankard/Assets/Scripts/AI/AISpawner.cs [hunk 1]
```diff
@@ -55,6 +51,28 @@ private void SpawnAdventurer()
...
+            ColorGameObjects(adventurer.adventurerClass.clothingColors, adventurer.clothesToColor);
+            ColorGameObjects(adventurer.racePref.skinColors, adventurer.skin);
+
+            adventurerObject.transform.position = randomSpawnPoint;
+            adventurerObject.SetActive(true);
+        }
+
+        /*
         Adventurer adventurer = Instantiate(prefab, randomSpawnPoint, Quaternion.identity);
         adventurer.race = (Race)Random.Range(0, Enum.GetNames(typeof(Race)).Length);
```
Comment (Lunium) [inline behind-the-tankard/Assets/Scripts/AI/AISpawner.cs]: Comments

File: behind-the-tankard/Assets/Scripts/AI/AISpawner.cs [hunk 2]
```diff
@@ -55,6 +51,28 @@ private void SpawnAdventurer()
     {
         Vector3 randomSpawnPoint = _spawnPoints[Random.Range(0, _spawnPoints.Count)];
 
+        GameObject adventurerObject = GenericObjectPooler.SharedInstance.GetPooledObject();
```
Comment (Lunium) [inline behind-the-tankard/Assets/Scripts/AI/AISpawner.cs]: GenericObjectPooler can now only be used for adventurers if implemented this way.

File: behind-the-tankard/Assets/Scripts/Components/GenericObjectPooler.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class GenericObjectPooler : MonoBehaviour
+{
+    public static GenericObjectPooler SharedInstance;
```
Comment (Lunium) [inline behind-the-tankard/Assets/Scripts/Components/GenericObjectPooler.cs]: Make this non static, so there can be multiple implementations of this for other object pools

File: behind-the-tankard/Assets/Scripts/AI/AISpawner.cs (hunk 2 above)
Comment (TJ-Smits) [reply behind-the-tankard/Assets/Scripts/AI/AISpawner.cs]: Correct, improved

File: behind-the-tankard/Assets/Scripts/Components/GenericObjectPooler.cs (hunk 3 above)
Comment (TJ-Smits) [reply behind-the-tankard/Assets/Scripts/Components/GenericObjectPooler.cs]: Correct, improved

File: behind-the-tankard/Assets/Scripts/AI/AISpawner.cs [hunk 4]
```diff
@@ -17,19 +17,21 @@ public class AISpawner : MonoBehaviour
...
     public Adventurer prefab;
 
-    public delegate void OnSpawn(Adventurer a);
-
-    public event OnSpawn onSpawn;
-
     private List<Vector3> _spawnPoints = new List<Vector3>();
     private bool _isOn;
 
+    private ObjectPooler pooler;
```
Comment (MattijnP) [inline behind-the-tankard/Assets/Scripts/AI/AISpawner.cs]: Rename this property to start with an underscore

File: behind-the-tankard/Assets/Scripts/Components/ObjectPooler.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            if (underParent)
+            {
+                if (hierarchyParent != null)
+                {
+                    obj = Instantiate(objectsToPool[number], hierarchyParent);
+                } else
+                {
+                    obj = Instantiate(objectsToPool[number], transform);
+                }
+            } else
```
Comment (MattijnP) [inline behind-the-tankard/Assets/Scripts/Components/ObjectPooler.cs]: Move both else's to a new line like this:
```cs
}
else
{
```

File: behind-the-tankard/Assets/Scripts/Components/ObjectPooler.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        for (int i = 0; i < pooledObjects.Count; i++)
+        {
+            if (!pooledObjects[i].activeInHierarchy)
+            {
+                return pooledObjects[i];
+            }
+        }
+
+        if (shouldExpand)
```
Comment (MattijnP) [inline behind-the-tankard/Assets/Scripts/Components/ObjectPooler.cs]: Reduce nesting by inverting this if statement like this:
```cs
if (!shouldExpand)
{
    return null;
}

int number = Random.Range(0, objectsToPool.Length);
// ...
```
If the if or else contain a return statement you can refactor the code so there will be no else statement.

File: behind-the-tankard/Assets/Scripts/Components/ObjectPoolerEditor.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using UnityEditor;
+
+[CustomEditor(typeof(ObjectPooler))]
+public class ObjectPoolerEditor : Editor
```
Comment (MattijnP) [inline behind-the-tankard/Assets/Scripts/Components/ObjectPoolerEditor.cs]: Move this class to the `Scripts/Editor` folder.

**Pull request merged**